Datagrams

dr7ana opened the pull request:
Currently:
- datagram sending functional with no packet-splitting policy (see 007 test case)
- added template parameterization to Network::endpoint s.t. datagram support and packet splitting is enabled for all connections in/out of one endpoint
- ngtcp2 transport params populated by top level endpoint configurations
- 007 unit tests for setting and getting datagram parameters

TODO (maybe for a second PR?):
- send datagrams with packet-splitting
- tetris buffer

dr7ana committed:
Unit tests refactored

- no waits, all using promises and/or threads
- focused on using threading for specific functionalities (ex: open stream)
- removed deprecated submodules leftover from merging PRs

dr7ana committed:
Test cleanups (+ optimizations and restructurings)

- Rebase onto warnings PR, streamlined implemented tests
- Replaced suboptimal mersenne twister
- Moved independant types to their own files
- Cleaned out utils.hpp, moved miscellaneous types to `types.hpp`
- No more '_t' names

dr7ana committed:
Test Cases (+ misc optimizations)

- failure cases and type-specific test cases added
- tests 006 and 007 combined
- unnecessary `{Outbound,Inbound}Context` -> `IOContext`
- `config_t` -> `user_config`

dr7ana committed:
Datagram support commit

- added template parameterization to Network::endpoint s.t. datagram support and packet splitting is enabled for all connections in/out of one endpoint
- ngtcp2 transport params populated by top level endpoint configurations
- 007 unit tests for setting and getting datagram parameters
- Datagram transmission functional with 007 unit test
- Comment blocks explaining datagram implementation and motivation
- Streamlined ngtcp2_conn_ref linkage in connection creation by handling it all in connection constructor; circumvents a small bug where the conn_ref object created in Connection::setup_tls_session was not copying the correct internals

TODO:
- Packet-splitting policy
- tetris buffer

dr7ana committed:
Unit testing CI fixes

- Added datagram retransmit timer for occasional hanging when datagram transmission fails on local testing; no latency so sometimes it trips over its own feet
- Monotonically increasing ports for localhost testing s.t. no ports are re-used between tests; aiming to fix sporadic CI fails where addresses are already bound (ex: mac builds)
- Construction of gnutls credentials using empty string was segfaulting instead of throwing on mac M1 locally and in CI (test 001)
- Added tls promise to success case in 004; aiming to fix Debian sid (ARM64) CI fail
- All tests running locally on M1 as of this commit

jagerman commented on src/connection.cpp:
```diff
@@ -766,6 +782,19 @@ namespace oxen::quic
 
+        if (uconfig.datagram_support)
+        {
+            params.max_datagram_frame_size = 65535;
```
Deserves a comment along the lines of:
```
// This is effectively an "unlimited" value, which lets us accept any size that fits into a QUIC packet.
```
because otherwise I was confused by us not being able to accept anything remotely as large as this.

tewinget commented on include/quic/connection.hpp:
```diff
@@ -93,63 +143,146 @@ namespace oxen::quic
+        const int _max_streams{DEFAULT_MAX_BIDI_STREAMS};
+        const bool _datagrams_enabled{false};
+        const bool _packet_splitting{false};
+        const Splitting _policy{Splitting::NONE};
```
The variable name `_policy` here seems ambiguous.

dr7ana replied:
I dunno. the enum class name is pretty clear what it is, so renaming it to `_splitting_policy` seems redundant. Also it's private and the getter is `Splitting Connection::packet_splitting_policy()`

tewinget replied:
If the getter is used anywhere/everywhere it is referenced even within this class, that will be fine.  If not, we will later have something like:

```
// user datagram comes in
if (_policy)
{
    // split the datagram if needed
}
else
{
    // don't split
}
```

In that case, reading the code some time in the future it could be confusing/ambiguous what `_policy` means.  It would likely become clear quickly enough by reading the body of the `if` statement, but that feels suboptimal.

dr7ana replied:
That's a fair point, ill go through and make sure that it uses the getter even internally

jagerman commented on src/connection.cpp:
```diff
@@ -766,6 +782,19 @@ namespace oxen::quic
+        if (uconfig.datagram_support)
+        {
+            params.max_datagram_frame_size = 65535;
+            // default ngtcp2 values set by ngtcp2_settings_default_versioned
```
If these are already set to these values by our calls to `ngtcp2_transport_params_default`/`ngtcp2_settings_default` then why set them again here (e.g. if the ngtcp2 defaults change, are we intentionally not willing to accept that change for some reason?)

dr7ana replied:
The only one i saw that `ngtcp2_settings_default` assigns is 

`settings->max_tx_udp_payload_size = 1500 - 48;`

I set it again explicitly in case we find the given overhead they estimated was insufficient and need to change the value, but I can absolutely remove it if its not helpful

The other one that `ngtcp2_transport_params_default` sets 

`params->max_udp_payload_size = NGTCP2_DEFAULT_MAX_RECV_UDP_PAYLOAD_SIZE;`

I set it to 65535 as per the rfc; their macro is 65527

tewinget commented on include/quic/connection.hpp:
```diff
@@ -93,63 +143,146 @@ namespace oxen::quic
+        const Splitting _policy{Splitting::NONE};
+        datagram_recv_callback dgram_data_cb;
+
+        /// Datagram Numbering:
```
I like how this is laid out.  I'd say that `packet` and `datagram` could be used more precisely here though, to avoid confusion.  For example, I'd say it should be phrased that we might split "packets" into multiple (two) datagrams.  Trying to cover all cases:

-----

> Each datagram ID is incremented by four from the previous one, regardless of whether we are splitting ~~datagrams~~ **packets**.

> In the unsplit packet scheme, the dgram ID of each ~~packet~~ **datagram** satisfies the rule:

> In the split-packet scheme, the dgram ID of the first of two ~~packets~~ **datagrams** satisfies the rule:

> The second of the two ~~packets~~ **datagrams** satisfies the rule:

> As a result, a packet-splitting endpoint should never send or receive a ~~packet~~ **datagram** whose ID is a perfect multiple of 4

dr7ana replied:
Thank you for the edits! That all makes perfect sense, I'll make the change

jagerman commented on src/address.cpp:
```diff
@@ -46,7 +46,7 @@ namespace oxen::quic
     std::string Path::to_string() const
     {
-        return "{{{} ➙ {}}}"_format(local, remote);
+        return "{{{} ➙ {}}}"_format(local.to_string(), remote.to_string());
```
These should not be necessary; local and remote should already be formattable.

jagerman commented on src/connection.cpp:
```diff
@@ -303,23 +339,27 @@ namespace oxen::quic
         sent_counter += n_packets;
 
-        auto rv = endpoint().send_packets(remote(), send_buffer.data(), send_buffer_size.data(), send_ecn, n_packets);
+        auto rv = endpoint().send_packets(_path.remote, send_buffer.data(), send_buffer_size.data(), send_ecn, n_packets);
```
Why make this change?

tewinget commented on include/quic/context.hpp:
```diff
@@ -12,73 +12,52 @@
-        void handle_inbound_opt(stream_data_callback_t func);
-        void handle_inbound_opt(stream_open_callback_t func);
-        void handle_inbound_opt(stream_close_callback_t func);
+        void handle_ioctx_opt(std::shared_ptr<TLSCreds> tls);
```
I think this one should be explicit from `Endpoint` (as you did with `Direction`) and the Endpoint should require a non-null `TLSCreds` on construction.  i.e. an Endpoint should have one `TLSCreds` shared among all connections, inbound or outbound.

dr7ana replied:
makes sense to me :+1:

jagerman reviewed (changes requested):
Overall comments:

Datagrams should use `NGTCP2_WRITE_DATAGRAM_FLAG_MORE` to allow packing small datagrams together into the same quic packet.  But this dovetails into the next point:

We don't want a separate loop & timer here for datagrams and stream data: we want, when feasible, to pack them together (e.g. if we have 100 bytes of stream data, and three 50-byte datagrams then we can fit all 4 things into one packet).

This ought to be not too difficult to implement with the current loop in `flush_streams` (which maybe we should rename to `flush_packets`):

- instead of appending one `nullptr` into the `strs` list we append *two* nullptrs.
- drop the `assert(stream || strs.empty())`
- inside the loop, if we get a nullptr and `strs` is *not* empty then it's the datagrams case (otherwise `strs` is empty so it's the final `-1` psuedo-stream case).
- for the datagrams case we need to skip if we have no datagrams to send, equivalent to what we do when a stream has no data to send in the `else if (bufs.empty()) { ... }` condition.
- if we're in the datagrams case then we need to call `ngtcp2_conn_writev_datagram` instead of `ngtcp2_conn_writev_stream`.  Most of the arguments will be the same, aside from the `ndatalen` and `stream_id` (which will be a `paccepted` and 0 for datagrams).
- the datagram return value is a bit wonky: unlike writev_stream, it can return a >0 final packet size but with `*paccepted = 0`, which indicates that the next datagram cannot be packed into the current packet: we need to fire off the just-finished packet, but then need to invoke the datagram handler again.
    - The easiest way to do this would be to re-add the `nullptr` stream pointer to the *beginning* of `strs` via a `push_front(nullptr)` instead of just-before-the-last, to avoid needing a duplicate writev_datagrams call, and then `continue;` the loop.
- At the end of the loop the current `if (!stream)` needs to be changed to `if (!stream && strs.empty())` but also we need the `else if` case right after than to also handle a "we have more datagrams" condition for re-appending `streams` to `strs`.

tewinget commented on include/quic/gnutls_crypto.hpp:
```diff
@@ -104,10 +129,10 @@ namespace oxen::quic
         ~GNUTLSSession();
 
-        void* get_session() override { return session; };
+        void* get_session() override { return reinterpret_cast<gnutls_session_t>(session); };
```
`session` is already a `gnutls_session_t`

dr7ana replied:
oops, was checking something debugging a CI build. thanks

tewinget replied:
looks like this was marked resolved here but the relevant change wasn't committed.

dr7ana replied:
sorry I clicked resolve on these as we were chatting about them on mumble about it (and some of your later comments), it should be changed

tewinget commented on include/quic/stream.hpp:
```diff
@@ -107,6 +85,44 @@ namespace oxen::quic
+            return sum;
+        }
+
+        inline size_t unacked() const { return unacked_size; }
```
We could expose this and `unsent` to the user I think.  They're const so it can't hurt, and I can see utility in the user having that information.

jagerman commented:
Also for sending datagrams, we want some feedback for the user in case the connection is congested so that they can (potentially) back off packet transmission (they could just ignore it, but it would be nice for them to be *able* to do something if they want).

Perhaps this might accomplish that:
- We keep a buffer of accepted datagrams.  Once we accept such a datagram then we'll send that datagram as soon as we can.
- We add a `private: std::atomic<bool> congested_;` along with a `public: bool is_congested() const { return congested_; }`
- Inside `flush_packets()` we do two things:
    - just after `if (n_packets > 0)` case, we set `congested_` to false
    - on any exits from the function that happen specifically because of congestion, we set `congested_ = true;` and drop all pending, queued datagrams.
- In the (application-facing) call to send a datagram, we check `is_congested()` and, if set, we immediately return a failure back to the user without queuing anything to the logic thread.  If not congested, we queue it.

Tom's suggestion:
- We add a configurable "max_pending_datagrams" settings on the connection (it could probably just be a public member, defaulting to 0), and then we make `send_datagram` accept if either `!congested_` *or* the current datagram queue is less than `max_pending_datagrams`.  Then instead of "drop all pending datagrams" on congestion we would only drop (from the *beginning* of the queue) as many as needed to lower the max pending datagrams to `max_pending_datagrams`.

tewinget commented on include/quic/utils.hpp:
```diff
@@ -37,28 +40,15 @@ extern "C"
     using namespace oxen::log::literals;
     using bstring = std::basic_string<std::byte>;
     using bstring_view = std::basic_string_view<std::byte>;
+    using buffer_que = std::deque<std::pair<bstring_view, std::shared_ptr<void>>>;
```
`buffer_queue` or `buffer_deque` is probably better.

tewinget commented on src/connection.cpp:
```diff
@@ -64,7 +67,14 @@ namespace oxen::quic
+    {
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
+        return static_cast<Connection*>(user_data)->recv_datagram(
+                {reinterpret_cast<const std::byte*>(data), datalen}, flags & NGTCP2_STREAM_DATA_FLAG_FIN);
```
Receive datagram has no `FIN` flag.  It seems to only have one flag, `0RTT`, which we don't support so don't care about.

The recv_datagram function below need not take a `bool fin` for the same reason.

tewinget commented on src/endpoint.cpp:
```diff
@@ -454,4 +484,9 @@ namespace oxen::quic
         return nullptr;
     }
 
+    bool Endpoint::in_event_loop() const
```
I don't see this used anywhere (could always be github trolling).

tewinget commented on tests/main.cpp:
```diff
@@ -7,7 +7,7 @@ int main(int argc, char* argv[])
 
     using namespace Catch::Clara;
-    std::string log_level = "debug", log_file = "stderr";
+    std::string log_level = "trace", log_file = "stderr";
```
idk if we want to default this to `trace` since it can easily be specified in the CLI.

jagerman committed:
Fix ci label

This build is on i386, not amd64.

dr7ana committed:
Improved Datagrams:

- introduced IOChannel abstract class inherited from by stream and datagramIO objects, allowing them to overlap a non-trivial amount of code
- datagrams and streams combined in one logic loop operating on one timer
- streams and datagrams use the same callbacks now

tewinget commented on include/quic/connection.hpp:
```diff
@@ -288,6 +274,10 @@ namespace oxen::quic
         std::map<int64_t, std::shared_ptr<Stream>> streams;
+        // datagram "pseudo-stream"
+        std::shared_ptr<DatagramIO> datagrams;
+        // stream "pseudo-stream" to finish off packets
```
to finish off packets, yes, but also for ngtcp2 to create any packets it may need us to send like ACKs, retransmits, etc.

tewinget replied:
Would be sufficient to just say "to represent ngtcp2 stream ID -1" though

tewinget commented on include/quic/datagram.hpp:
```diff
@@ -0,0 +1,95 @@
+        size_t unsent() const override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return std::numeric_limits<size_t>::min();
```
Feels like this should have an actual value it could return.

dr7ana replied:
For sure, I just needed a placeholder here in case it got accidentally called that wasn't in the realm of a realistic response

tewinget commented on include/quic/utils.hpp:
```diff
@@ -108,6 +108,8 @@ namespace oxen::quic
     inline constexpr size_t MAX_LAZY_PMTUD_UDP_PAYLOAD = MAX_PMTUD_UDP_PAYLOAD;                // 1452
     inline constexpr size_t MAX_GREEDY_PMTUD_UDP_PAYLOAD = (MAX_LAZY_PMTUD_UDP_PAYLOAD << 1);  // 2904
 
+    inline constexpr size_t FLAG_WRITE_MORE = 0x01u;
```
This should be set to ngtcp2's constant.  I think for now we have to assume the datagram and stream constants will always equal one another, but they *could* change, and nevertheless it's better to know where the number came from than to have a magic number.

tewinget commented on src/connection.cpp:
```diff
@@ -514,119 +395,172 @@ namespace oxen::quic
             }
         }
 
+        size_t max_datagram_size = -1;
```
It shouldn't matter, but this should probably default to 0 rather than -1, as `size_t` is unsigned.

dr7ana replied:
Thanks that was a dumb oversight

jagerman commented on src/connection.cpp:
```diff
@@ -350,120 +395,172 @@ namespace oxen::quic
-                    strs.push_back(stream_ptr.get());
+                if (stream_ptr and not stream_ptr->_sent_fin)
+                    channels.push_back(stream_ptr.get());
             }
```
We should append the datagram channel right here rather than later, so that it also ends up randomly positioned in the initial channel list.

tewinget commented on src/connection.cpp:
```diff
@@ -514,119 +395,172 @@ namespace oxen::quic
+                        ts);
+
+                log::trace(log_cat, "add_stream_data for stream {} returned [{},{}]", stream_id, nwrite, ndatalen);
+                // ndatalen "should" be >0 for streams and <0 for psuedo streams
```
The docs seem sorta ambiguous, but it *looks* like for the `stream id == -1` case, ndatalen will be either -1 or 0, so not necessarily negative?

dr7ana replied:
I was unsure, i never saw a ndatalen of 0 for the stream id of -1, and like you said the docs were ambiguous. ill look into it :+1:

tewinget commented on src/connection.cpp:
```diff
@@ -514,119 +395,172 @@ namespace oxen::quic
+                }
+
+                // fatal error
+                if (nwrite < -500)
```
should call `ngtcp2_err_is_fatal(nwrite)` or w/e that func is named.

jagerman commented on src/connection.cpp:
```diff
@@ -350,120 +395,172 @@ namespace oxen::quic
         size_t stream_packets = 0;
-        while (!strs.empty())
-        {
+        ngtcp2_ssize nwrite = 0;
+        ngtcp2_ssize ndatalen;
+        int datagram_accepted = 0;
+        bool write_successful = false;
```
These variables should be inside the loop, just before used.  Otherwise their value can persist across loop iterations.

jagerman replied:
Perhaps rename to `packet_filled` or something along those lines?  (Unless I'm mistaken as to what this variable is meant to do)

jagerman commented on src/connection.cpp:
```diff
@@ -350,120 +395,172 @@ namespace oxen::quic
+                // both must be true to be successful
+                if (datagram_accepted != 0)
+                {
+                    // if the datagram was accepted at all, drop it from the buffer
```
```suggestion
                    // if the datagram was accepted, drop it from the buffer
```
The "at all" implies it could be partially accepted, but it cannot be (ngtcp2 guarantees this).

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -0,0 +1,95 @@
+        virtual void set_fin(bool v) = 0;
+        virtual size_t unsent() const = 0;
+        virtual void wrote(size_t n) = 0;
+    };
```
This class is useful internally for managing the buffers, but it feels like this internal implementation detail is leaking out to the libquic-using application and forcing complications on *outside* code to make *inside* code more maintainable.

From an application's point of view, I'm not sure what the value is in having Stream and DatagramIO inherit from a base type (and then giving me the base type): that just means the application has to untangle the object structure to get at the thing they care about (which will be a stream *or* a datagram, but rarely or perhaps never both).  And while we can duplicate most of Stream's public interface into IOChannel (as is done here) that is an awkward API for the caller when using datagrams because we are going through a base class that has a bunch of methods that are meaningless for a datagram channel.

If you were to use the base class *internally* but still have the user use Stream and DatagramIO, we would get more manageable code internally but also more manageable code externally, which seems like a win.

dr7ana replied:
That's a really good point that I didn't consider. I guess I was trying to do 3 different things at the same time, and it "almost" worked
1) introduce the base IOChannel class and have datagram and stream inherit from them
2) combine the error handling for flushing packets into one efficient loop
3) have the user-provided callbacks interoperate with streams without issue

I'll untangle this, starting probably with point 2

jagerman commented on include/quic/endpoint.hpp:
```diff
@@ -53,7 +47,14 @@ namespace oxen::quic
         Endpoint& operator=(Endpoint&&) = delete;
 
-        explicit Endpoint(Network& n, const Address& listen_addr);
+        std::string local_addr() { return _local.to_string(); }
```
I think this method was added for debug, but isn't really needed: if someone wants it as a string they have a few options: `ep.local().to_string()` or `"{}"_format(ep.local())`, or even just pass `ep.local()` if it's already going into a formattable string.

dr7ana replied:
Sorry just saw this, will fix and rebase

dr7ana committed:
Fix for address handling and dual-stack socket addressing

- All tests use kernel chosen hosts and ports
- Binding a server to a default ipv6 address [::]:0 and connecting to a client on localhost results in dropped packts, as the return path uses the ipv4 mapped ipv6 address `::ffff:127.0.0.1:<port>`. As a result, the packets get dropped by the client
- Added test case to 001 verifying ipv6 as remote address
- Added cli arg "--no-ipv6" that deactivates it for CI
- Misc unit test cleanups
- Close #23

tewinget commented on include/quic/endpoint.hpp:
```diff
@@ -90,12 +90,9 @@ namespace oxen::quic
             auto f = p.get_future();
 
-            Path _path;
+            auto& _r = (_local.is_ipv6()) ? Address::map_ipv4_as_ipv6(remote) : remote;
```
What happens if `remote` is also/already ipv6?

jagerman replied:
This is making a reference to a temporary.

It would be cleaner to have the mapping function as a mutator of an Address rather than a static method making a new copy of it (because it seems unlikely that where this would be used would ever care about keeping the original).

jagerman replied:
> What happens if `remote` is also/already ipv6?

Then you'll get an error as soon as you try to send anything.  I suppose that error could be detected earlier and throw, though.

dr7ana replied:
Yeah I need to account for that error

@jagerman I made it into a static method because the Address passed to connect is a const reference, so I would have had to const cast it

dr7ana replied:
fixed and added a test case

tewinget reviewed (approved):
Seems fine to me.

jagerman commented on include/quic/endpoint.hpp:
```diff
@@ -83,28 +85,29 @@ namespace oxen::quic
         // creates new outbound connection to remote; emplaces conn/interface pair in outbound map
         template <typename... Opt>
-        std::shared_ptr<connection_interface> connect(const Address& remote, Opt&&... opts)
+        std::shared_ptr<connection_interface> connect(Address& remote, Opt&&... opts)
```
```suggestion
        std::shared_ptr<connection_interface> connect(Address remote, Opt&&... opts)
```
so that it can be called with a const value (it means it gets passed by copy, but that's not a big deal).

dr7ana replied:
will commit separately :+1:

jagerman commented on include/quic/endpoint.hpp:
```diff
@@ -83,28 +85,29 @@ namespace oxen::quic
             auto f = p.get_future();
 
-            net.call([&opts..., &p, path = Path{local, remote}, this]() mutable {
+            if (_local.is_ipv6() && not remote.is_ipv6())
```
```suggestion
            if (_local.is_ipv6() && !remote.is_ipv6())
```

dr7ana replied:
same as above
